### DISCUSSION TITLE
How to install hbase through paketo buildpacks

### DISCUSSION THREAD

**Discussion by AshwinSridharan0410** [paketo-buildpacks/feedback#264] [Java Team]
I want to replicate this Dockerfile . Is it possible to do in paketo buildpacks
**FROM openjdk:17-jdk
ENV HBASE_HOME='/app/hbase-2.1.6' \
    PATH="${JAVA_HOME}/lib:/app/hbase-2.1.6/bin:${PATH}"
WORKDIR /app
COPY hbase-2.1.6-bin.tar.gz .
RUN tar -xzf hbase-2.1.6-bin.tar.gz
COPY ./hbase-site*.xml hbase-2.1.6/conf/
COPY target/*.jar alarm.jar
COPY docker-entrypoint.sh /
RUN chmod +x /docker-entrypoint.sh
EXPOSE 80
CMD ["/bin/sh", "-c", "/docker-entrypoint.sh"]**

The copying of files can be done by using BP_INCLUDE_FILES and to set the docker-entrypoint shell file as the entrypoint, I will add a Procfile and set the start command accordingly. But how to install hbase without using Procfile(curl command ) or adding it in my stack ?

**Comment by dmikusa** [comment]
The primary target for buildpacks is to take your source code and turn that into a container image. It excels when working with an application that you're developing, but in some cases, you can use it for packaging up 3rd party apps. It's not meant as a complete replacement for Dockerfiles.

This falls into the 3rd party app category and that may or may not work, depending on how the app you're trying to deploy is built.

If you wanted to do this, the workflow would look something like this:

1. Download the app.
2. Build one of the following: a WAR file, an executable JAR file, or a dist-zip archive from the app. If one is made available by the 3rd party, that works too.
3. Run `pack build -p path/to/artifact-from-step-2`. This will run buildpacks with the artifact and if all goes well produce an image you can use.

The reason I specifically mention WAR, executable JAR, or dist-zip file is that that's what the Java buildpack knows how to deploy. If your 3rd party app is run in a different way, then you've got more work to do. You'd either need to get it to one of the supported deployment options or you'd need to use Procfile buildpack and specify a start command manually. I can't see what's being done in the `docker-entrypoint.sh` file, but if it's running `java -jar` then you have an executable JAR and you should be OK.

Other areas that can cause issues with 3rd party apps:

- Apps that write stuff to the local file system. Buildpacks has pretty restrictive permissions by default, for good security. This can cause issues. Also, local file systems are ephemeral so if you need to do that, you might need a volume mount when you run the app.
- Apps that need to do stuff as root. Buildpacks don't run as root nor do the containers that get produced. This is not an option.
- Missing configuration files or assumed file structures that don't exist. If the app expects files to exist in certain locations, you may need to reconfigure the app to pull configuration from somewhere else. That or map those files into the container at the expected locations.

Hope that helps!

**Comment by AshwinSridharan0410** [comment]
Thanks for the reply @dmikusa .

The application I am trying to deploy is a JAR file. And what's happening inside docker-entrypoint.sh 

docker-entrypoint.sh 
**#!/bin/bash
[[ -z $Environment ]] && echo "No Environment variable is specified! Exit." && exit 0
cd /app/hbase-2.1.6/conf/
[[ ! -z $Environment && "$Environment" == "DEV"* ]] && \
echo "Using hbase-site-dev.xml" && \
ln -sf hbase-site-dev.xml hbase-site.xml
cd /app
java --add-opens=java.base/java.nio=ALL-UNNAMED -javaagent:dd-java-agent.jar -jar alarm.jar**

Ideally the approach that I want to do is to run my jar file in a hbase environment and while keeping the start command to docker-entrypoint.sh, it is checking for the env which will be passed when deploying as a (volume mount -> env.variable) and run the jar file along with the xml files we are copying in the docker image.

For this , I had two approaches -> 
1) Use our java buildpacks and have a procfile which does the logic along with passing some env. variables (BP_INCLUDE_FILES)
2) Or write a custom buildpack (build and detect script) -> In the build script, install hbase and copy the xml files and set the launch command to the shell script (docker-entrypoint.sh)

Hope I am clear. Can you tell me the exact approach that I need to do in order to build it through paketo buildpacks.Please provide your views.

Thanks !

**Comment by AshwinSridharan0410** [comment]
@dmikusa , any comments here for this ?

**Reply by dmikusa** [reply]
Sorry, nothing beyond what I originally posted. I don't know the tool you're trying to install so I can't really say what would be the best approach. Again, buildpacks work great for turning app source code into containers. When you get into using them with "someone else's app", it can be tricky. You can sometimes do it, but if the friction becomes too much, then you might want to look at writing a buildpack like you said. You have more flexibility that way. I'll give you an example. If I want to deploy Tomcat, I'm not going to do that from source code. It's a prepackaged app, designed to be run in a certain way so writing a buildpack makes sense here. You have more flexibility to install and configure Tomcat. Again, I don't know the tool you're writing about but maybe that can help you decide on the approach to take.

**Reply by AshwinSridharan0410** [reply]
Yep , thanks for the suggestion @dmikusa
